File: react-js/src/Game.js
import React from "react";
import {calculateWinner} from "./Helper";
import Board from "./Board";

class Game extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      xIsNext: true,
      stepNumber: 0,
      history: [{
        squares: Array(9).fill(null),
      }],
      showPlayerSelection: true,
    };
  }

  handleClick(i) {
    const newHistory = this.state.history.slice(0, this.state.stepNumber + 1);
    const current = newHistory[newHistory.length - 1];
    const squares = current.squares.slice();

    if (calculateWinner(squares) || squares[i]) {
      return;
    }

    squares[i] = this.state.xIsNext ? 'X' : 'O';
    this.setState({
      xIsNext: !this.state.xIsNext,
      stepNumber: newHistory.length,
      history: newHistory.concat([{
        squares: squares,
      }]),
      showPlayerSelection: false,
    });
  }

  jumpTo(step) {
    this.setState({
      stepNumber: step,
      xIsNext: (step % 2) === 0,
    });
  }

  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    let isStepLeft = true;
    const moves = history.map((step, move) => {
      isStepLeft = step.squares.some(square => square === null);
      const desc = move ? 'Go to move #' + move : 'Go to game start';
      return (
          <li key={move}>
            <button onClick={() => this.jumpTo(move)}>{desc}</button>
          </li>
      );
    });

    let status;
    if (winner) {
      status = 'Winner: ' + winner;
    } else if (isStepLeft) {
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
    } else {
      status = 'Nobody won :(';
    }

    return (
        <div className="game">
          {this.state.showPlayerSelection && <div className="players">
            <button onClick={() => {
              this.setState({
                xIsNext: true,
                showPlayerSelection: false,
              });
            }}>Player X
            </button>
            <button onClick={() => {
              this.setState({
                xIsNext: false,
                showPlayerSelection: false,
              });
            }}>Player O
            </button>
          </div>}
          <div className="game-board">
            <Board
                squares={current.squares}
                onClick={(i) => this.handleClick(i)}
            />
          </div>
          <div className="game-info">
            <div>{status}</div>
            {(winner || !isStepLeft) && <button onClick={() => {
              this.jumpTo(0);
              this.setState({
                history: [{
                  squares: Array(9).fill(null),
                }],
              });
            }}>Start new game</button>}
            <ol>{moves}</ol>
          </div>
        </div>
    );
  }
}

export default Game;

File: kotlin-multiplatform/src/jsMain/kotlin/Game.kt
import de.rahn.kotlin.multiplatform.tictactoe.History
import de.rahn.kotlin.multiplatform.tictactoe.SquareType
import de.rahn.kotlin.multiplatform.tictactoe.calculateWinner
import react.Props
import react.dom.attrs
import react.dom.button
import react.dom.div
import react.dom.events.MouseEvent
import react.dom.li
import react.dom.ol
import react.dom.onClick
import react.fc
import react.useState

val game = fc<Props> {
  val initialHistory by useState(arrayOf(History()))

  var xIsNext by useState(true)
  var stepNumber by useState(0)
  var history by useState(initialHistory)
  var showPlayerSelection by useState(true)

  fun handleClick(i: Int) {
    val newHistory = history.slice(0..stepNumber).toMutableList()
    val current = newHistory.last()
    val squares = current.squares.copyOf()

    if (calculateWinner(squares) !== null || squares[i] !== null) {
      return
    }

    squares[i] = if (xIsNext) SquareType.X else SquareType.O
    newHistory.add(History(squares))
    xIsNext = !xIsNext
    stepNumber = newHistory.size - 1
    history = newHistory.toTypedArray()
    showPlayerSelection = false
  }

  fun jumpTo(step: Int) {
    stepNumber = step
    xIsNext = step % 2 == 0
  }

  val current = history[stepNumber]
  val winner = calculateWinner(current.squares)

  var isStepLeft = true
  val moves = history.mapIndexed { move: Int, step: History ->
    isStepLeft = step.squares.any { it === null }
    val desc = if (move != 0) "Go to move #$move" else "Go to game start"
    Pair(move, desc)
  }

  val status = when {
    winner is SquareType -> "Winner: ${winner.name}"
    isStepLeft -> "Next player: ${if (xIsNext) SquareType.X else SquareType.O}"
    else -> "Nobody won :("
  }

  div(classes = "game") {
    if (showPlayerSelection) {
      div(classes = "players") {
        button {
          attrs {
            onClick = {
              xIsNext = true
              showPlayerSelection = false
            }

            +SquareType.X.name
          }
        }
        button {
          attrs {
            onClick = {
              xIsNext = false
              showPlayerSelection = false
            }

            +SquareType.O.name
          }
        }
      }
    }
    div(classes = "game-board") {
      child(board) {
        attrs {
          squares = current.squares
          onClick = { i -> handleClick(i) }
        }
      }
    }
    div(classes = "game-info") {
      div {
        +status
      }
      if (winner is SquareType || !isStepLeft) {
        button {
          attrs {
            onClick = {
              jumpTo(0)
              history = initialHistory
              showPlayerSelection = true
            }

            +"Start new game"
          }
        }
      }
      ol {
        for ((move, desc) in moves) {
          li {
            key = "$move"

            button {
              attrs {
                onClick = { jumpTo(move) }
              }

              +desc
            }
          }
        }
      }
    }
  }
}
